### PR TITLE
Add Procedure for Triggering DAG Runs Across Deployments

### LINKED ISSUE #1388: Add steps on making API requests across Deployments
### Context

Rappi is an Astronomer Software customer that is migrating to Astro this year. As part of that migration, they have a use case where they need to trigger DAGs cross-deployments. This requires calling the Airflow API.

Our CRE team has recently found that we can support this workflow, but it's tricky to set up as an extension of our current functionality.

### Ask

In our Airflow API doc (https://docs.astronomer.io/astro/airflow-api):

- Create a new topic for `API Request across Deployments`
- Include 1-2 sentences on "why" you might want to do this
- State that this works for any Deployment in any cluster or Workspace on Astro, given that the API key is securing the authentication
- Add steps required for this to work
     - 1. Create a Deployment API key for Deployment A
     - 2. Store the Deployment API key in Deployment B (not sure how)
     - 3. Then, in each task in Deployment B that requires using the Airflow API, use the API key from Deployment A to call the OAuth handshake:
         
          ```
          curl --location --request POST "https://auth.astronomer.io/oauth/token" \
                 --header "content-type: application/json" \
                 --data-raw '{
                     "client_id": "<api-key-id>",
                     "client_secret": "<api-key-secret>",
                     "audience": "astronomer-ee",
                     "grant_type": "client_credentials"}'
          ```
    - 4. Then, use the token in the response to actually make the API call

Google Doc from @collin-astro: https://docs.google.com/document/d/1P6y2DuHu_OmdhAJx1Qac7Cqf9XD9RkgJr5VYzZ1ysYE/edit

### PR DESCRIPTION
**Pull request by kwcanuck**
Resolves #1388

### CODE REVIEW

**Comment by kwcanuck** [comment]
@collin-astro there are a number of gaps in the preliminary draft of the procedure I prepared. I'm hoping you can help answer the following questions:

- Why would a user want to complete this task? What benefit does it provide?

- How does a user store the access token for one Deployment in another Deployment?

- What does a user need to do to "make the Airflow API call"? Can we be more specific, or provide an example?

- Are specific permissions required to perform this procedure?

**Comment by collin-astro** [comment]
I believe this is meant for @collinmcnulty

**Comment by collinmcnulty** [comment]
You actually can't do steps 1 and 2. Step 3 will spit out the access token as part of a json, which you must rip out and store as a string. You have to store the api key and secret api key on Deployment B.

In this wording, it's deployment B triggering a dag in deployment A. I've been using "triggering deployment" and "target deployment" colloquially.

**Comment by collinmcnulty** [comment]
A user may want to do this if they are (as we suggest) using different deployments to cordon off different teams workflows. This would only work from an organization mgmt perspective though, as once you have an api key, you've hosed the security barrier between the two (or at least made it one way only).

**Comment by collinmcnulty** [comment]
I think how to "make the Airflow API call" is covered adequately above in this same document.

File: astro/airflow-api.md [hunk 1]
```diff
@@ -198,3 +198,50 @@ response = requests.patch(
 print(response.json())
 # Prints data about the DAG with id <dag-id>
 ```
+## Trigger DAG runs across Deployments
+
+When you need to trigger DAGs in multiple Deployments, you can use the Airflow API. Triggering DAGS in multiple Deployments is sometimes necessary when you're using multiple Deployments to separate team workflows.
+
+This methodology works with any Deployment in any Astro Workspace or cluster. 
+
+1. On the target Deployment, create an API key ID and API key secret. See [Set environment variables on Astro](api-keys.md#create-an-api-key).
```
Comment (collinmcnulty) [inline astro/airflow-api.md:207]: Wrong link title?

File: astro/airflow-api.md [hunk 2]
```diff
@@ -198,3 +198,50 @@ response = requests.patch(
 print(response.json())
 # Prints data about the DAG with id <dag-id>
 ```
+## Trigger DAG runs across Deployments
+
+When you need to trigger DAGs in multiple Deployments, you can use the Airflow API. Triggering DAGS in multiple Deployments is sometimes necessary when you're using multiple Deployments to separate team workflows.
```
Comment (collinmcnulty) [inline astro/airflow-api.md:203]: I'd add "Similar DAGs can be created for any operation supported in the Airflow API"

**Review by collinmcnulty — CHANGES_REQUESTED** [review]

**Review by collinmcnulty — APPROVED** [review]

File: astro/airflow-api.md (hunk 2 above)
Comment (paolaperaza) [inline astro/airflow-api.md:203]:
```suggestion
You can use the Airflow REST API to make a request in one Deployment that triggers a DAG run in a different Deployment. This is sometimes necessary when you have interdependent workflows across multiple Deployments. On Astro, you can do this for any Deployment in any Workspace or cluster. 
```

File: astro/airflow-api.md [hunk 3]
```diff
@@ -198,3 +198,50 @@ response = requests.patch(
...
+
+This methodology works with any Deployment in any Astro Workspace or cluster. DAGs similar to the one provided can be created for any operation supported by the Airflow API.
```
Comment (paolaperaza) [inline astro/airflow-api.md:205]:
```suggestion
This topic has guidelines on how to trigger a DAG run, but you can modify the example DAG provided to trigger any request that's supported in the Airflow REST API.
```

File: astro/airflow-api.md [hunk 4]
```diff
@@ -198,3 +198,50 @@ response = requests.patch(
...
+        start_date=datetime(2022, 1, 1),
+        catchup=False)
+    def triggering_dag():
+        @task
+        def get_token():
+            response = requests.post("https://auth.astronomer.io/oauth/token",
+                                    json={"audience": "astronomer-ee",
+                                        "grant_type": "client_credentials",
+                                        "client_id": KEY_ID,
+                                        "client_secret": KEY_SECRET})
```
Comment (paolaperaza) [inline astro/airflow-api.md:231]:
```suggestion
                                        "client_id": {KEY_ID},
                                        "client_secret": {KEY_SECRET})
```

pretty sure these are needed?

File: astro/airflow-api.md [hunk 5]
```diff
@@ -198,3 +198,50 @@ response = requests.patch(
...
+
+When you need to trigger DAGs in multiple Deployments, you can use the Airflow API. Triggering DAGS in multiple Deployments is sometimes necessary when you're using multiple Deployments to separate team workflows.
+
+This methodology works with any Deployment in any Astro Workspace or cluster. DAGs similar to the one provided can be created for any operation supported by the Airflow API.
+
+1. On the target Deployment, create an API key ID and API key secret. See [Create an API key](api-keys.md#create-an-api-key).
+
+2. On the trigger Deployment, save the API key ID and API key secret as `KEY_ID` and `KEY_SECRET` environment variables. Make `KEY_SECRET` secret. See [Set environment variables on Astro](environment-variables.md).
+
+3. Create and then run the following DAG to start a DAG run on the target Deployment from the triggering Deployment:
```
Comment (paolaperaza) [inline astro/airflow-api.md:211]:
```suggestion
3. In your DAG, write a task called `get-token` that uses your Deployment API key ID and secret to make a request to the Astronomer API that retrieves the access token that's required for authentication to the Airflow API. In another task called `trigger_external_dag`, use the access token to make a request to the `dagRuns` endpoint of the Airflow REST API. Make sure to replace `<target-deployment-url>` with your own value.

Example:
```

File: astro/airflow-api.md [hunk 6]
```diff
@@ -198,3 +198,50 @@ response = requests.patch(
...
 ```
+## Trigger DAG runs across Deployments
+
+When you need to trigger DAGs in multiple Deployments, you can use the Airflow API. Triggering DAGS in multiple Deployments is sometimes necessary when you're using multiple Deployments to separate team workflows.
+
+This methodology works with any Deployment in any Astro Workspace or cluster. DAGs similar to the one provided can be created for any operation supported by the Airflow API.
+
+1. On the target Deployment, create an API key ID and API key secret. See [Create an API key](api-keys.md#create-an-api-key).
+
+2. On the trigger Deployment, save the API key ID and API key secret as `KEY_ID` and `KEY_SECRET` environment variables. Make `KEY_SECRET` secret. See [Set environment variables on Astro](environment-variables.md).
```
Comment (paolaperaza) [inline astro/airflow-api.md:209]:
```suggestion
2. On the triggering Deployment, set the API key ID and API key secret from the target Deployment as `KEY_ID` and `KEY_SECRET` environment variables in the Cloud UI. Make `KEY_SECRET` secret. See [Set environment variables on Astro](environment-variables.md).
```

**Review by paolaperaza — COMMENTED** [review]
Quick comment here @kwcanuck -- TL;DR I don't think we can frame step 3 as "copy paste this!" given that there are other things they'll want to customize.

Added more context to what the DAG is doing (feel free to organize differently) and framed it explicitly as an "example"

**Review by paolaperaza — APPROVED** [review]

File: astro/airflow-api.md [hunk 7]
```diff
@@ -198,3 +198,50 @@ response = requests.patch(
...
+
+3. In your DAG, write a task called `get-token` that uses your Deployment API key ID and secret to make a request to the Astronomer API that retrieves the access token that's required for authentication to the Airflow API. In another task called `trigger_external_dag`, use the access token to make a request to the `dagRuns` endpoint of the Airflow REST API. Make sure to replace `<target-deployment-url>` with your own value. For example:
+
+    ````python
+    import requests
+    from datetime import datetime
+    import os
+    from airflow.decorators import dag, task
+    KEY_ID = os.environ.get("KEY_ID")
+    KEY_SECRET = os.environ.get("KEY_SECRET")
```
Comment (pgagnon) [inline astro/airflow-api.md:219]: We should use Airflow variables instead of env vars here.

File: astro/airflow-api.md [hunk 8]
```diff
@@ -198,3 +198,50 @@ response = requests.patch(
...
+        catchup=False)
+    def triggering_dag():
+        @task
+        def get_token():
+            response = requests.post("https://auth.astronomer.io/oauth/token",
+                                    json={"audience": "astronomer-ee",
+                                        "grant_type": "client_credentials",
+                                        "client_id": {KEY_ID},
+                                        "client_secret": {KEY_SECRET})
+            return response.json()["access_token"]
```
Comment (pgagnon) [inline astro/airflow-api.md:232]: Passing the access token as an xcom is potentially insecure depending on the backend used.

File: astro/airflow-api.md (hunk 7 above)
Comment (kwcanuck) [reply astro/airflow-api.md:219]: @pgagnon thanks for the feedback. Can you provide the _specific_ edits you are requesting? What specific Airflow variables should be used?

This PR has been merged, so I'll need to open a new PR to address your concerns.

File: astro/airflow-api.md [hunk 9]
```diff
@@ -198,3 +198,50 @@ response = requests.patch(
...
+        @task
+        def trigger_external_dag(token):
+            dag_id = "target"
+            response = requests.post(
+                url=f"{AIRFLOW_URL}/api/v1/dags/{dag_id}/dagRuns",
+                headers={
+                    "Authorization": f"Bearer {token}",
+                    "Content-Type": "application/json"
+                },
+                data='{"logical_date": "' + datetime.utcnow().isoformat().split(".")[0] + 'Z"}'
```
[the remark this answers is not in the document]
Comment (kwcanuck) [reply astro/airflow-api.md:242]: @pgagnon again, can you specify the specific edit and location you are requesting? What specifically needs to be replaced with `strftime()`?

File: astro/airflow-api.md (hunk 8 above)
Comment (kwcanuck) [reply astro/airflow-api.md:232]: @pgagnon good to know. However, what do you want to tell users specifically? Should they be using this methodology?

**Pull request merged**